### DISCUSSION TITLE
Color + White

### DISCUSSION THREAD

**Discussion by hendrikBaechle** [HomeSpan/HomeSpan#354] [Q&A]
I want to control an rgbcct led strip (with red, green, blue, warm and cold white led. I’ve struggled to add upon the code from example 10 and wanted to ask if controlling that lightstrip with just a single tile is even possible. Thanks in advance.

**Comment by HomeSpan** [comment]
There are a number of ways to achieve this, ranging from the creation of Custom Characteristics to use with something like Eve for HomeKit, to creating generalized logic to map 3 Hue/Saturation/Brightness values to 5 LED Strip parameters (RGB + WW and CW).

However, I think the easiest and most flexible method that provides the most granular control would be to create a single Accessory with 2 Lightbulb Services - one with Hue/Saturation/Brightness, and one with ColorTemperature - as such:

```C++
  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify();

    new Service::LightBulb();
      new Characteristic::Name("RGB");
      new Characteristic::On();
      new Characteristic::Brightness();
      new Characteristic::Hue();      
      new Characteristic::Saturation();      
      
    new Service::LightBulb();
      new Characteristic::Name("White");
      new Characteristic::On();
      new Characteristic::Brightness();
      new Characteristic::ColorTemperature();
```

This will create a single tile with two separate controls.  The RGB light will provide for full color control in the Home App (as well as a "color temperature" tab, but the Home App just maps this back to Hue/Saturation/Brightness so it can be ignored).  The White light will provide the color temperature control.

You can control the lightstrip RGB inputs using the Hue/Saturation/Brightness values passed from the Home App for the first LightBulb, and control the lightstrip CW and WW inputs using the ColorTemperature/Brightness values passed from the Home App for the second LightBulb.

Note HomeSpan includes a specific class that supports NeoPixel RGBW LEDs but you'll need to write your own driver to control an RGBWW LED - what specific LED strip are you using?

**Comment by hendrikBaechle** [comment]
![ECC9D477-1E92-45BD-8CF5-6410B0778631](https://user-images.githubusercontent.com/110624867/183301656-65670dc3-5116-4040-bd9a-ef9a72988539.jpeg)
![5233B86E-F88D-4050-A75C-BD1EB79272D2](https://user-images.githubusercontent.com/110624867/183301662-6a73c134-d81a-48dd-8081-cf6346daaa47.jpeg)

I‘m using this analog 6-channel Led Strip from "IDERWEL".

I have so far just added the color characteristic to the same service as the hue, saturation, brightness and power state. That causes a bit of weird behaviour though. But at least I got it to work sometimes.

**Comment by HomeSpan** [comment]
Thanks for the details - looks like this is not an addressable pixel-style strip but simply a direct-drive (all colors on the strip are the same).  If so, you should be able to use the existing Example 10 to drive the RGB inputs and tie the CW and WW inputs to ground as a start.  Note you'll need some form of transistor logic or other type of driver to power this strip since it is designed for 12V.

Once that's working you can layer in logic for the CW and WW inputs using Example 10 as a model.  My recommendation would be to convert the ColorTemperature output from the Home App to two PWM signals where one drives CW to a maximum and sets WW off when the ColorTemp is set to the coolest color, and vice versa for when the colorTemp is set to the warmest color.  Note that the Home App outputs the inverse of color temp, ranging from 150-400.  Not sure why Apple decided to do this instead of simply outputting temperature settings directly.

**Comment by hendrikBaechle** [comment]
Yes, the LED Strip is not adressable (I thought that’s what analog means).
And I, of course can’t drive the LED Strip directly off of the ESP32, that why I built a test stand for now. (Pictures attached). I got the warm white and cold white LED‘s to light up when I use the color temperature panel in the home app.

The last two problems I need to fix are:

• When I ask Siri for warm/cold white, it uses the RGB and not the CW and WW LED‘s (don’t know if that’s fixable)
• The Home App doesn’t retain the color temperature (but sends the update to the ESP32) (Video attached)

https://user-images.githubusercontent.com/110624867/183428854-4c142f92-d517-4ae5-b30c-1338995aba40.MOV


![C4D78DC3-CF1A-4D70-B9B7-B8E1C7624A0F](https://user-images.githubusercontent.com/110624867/183426103-7d219517-ac60-42fb-a949-806d9642419e.jpeg)

![992751CF-FAA9-4DF0-AA3C-75E4BED269DA](https://user-images.githubusercontent.com/110624867/183426065-563b24fe-05c3-4670-8123-1d672729aedc.jpeg)

**Comment by HomeSpan** [comment]
Unfortunately I think you've encountered some of the many limitations of Siri and HomeKit.  I have found there is less I can do with Siri than I can directly with the Home App, and that the Home App itself does not always behave the way it should (in particular, valves and faucets often don't show up in the interface).

You may also want to try using Eve for HomeKit as a complement to Apple's Home App.  It's a free app and I find it a little more stable for some of the HomeKit Services.
